### PR TITLE
[Persistence] TreeStore Refactor

### PR DESCRIPTION
**Pull request by dylanlott**
## Description

Refactors the stateTrees component behind the TreeStore interface.

## Issue

Part of a refactor effort towards implementing #562

Part of a series of refactors in the persistence package, with #736 being the first.

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Pulls the stateTrees component down into its own component behind a dedicated interface for managing the SMT store.

## Testing

- [x] `make develop_test`; if any code changes were made
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: persistence/context.go
Comment (Olshansk) [inline persistence/context.go:34]: 🙌

File: shared/modules/persistence_module.go
Comment (Olshansk) [inline shared/modules/persistence_module.go:46]: Optional NIT: I feel like single lining this whole statement (even if it's > 80 chars) will make it easier to read

File: shared/modules/persistence_module.go
Comment (Olshansk) [inline shared/modules/persistence_module.go:41]: `by the treeStore` - What is the `treeStore`?

File: shared/modules/persistence_module.go
Comment (Olshansk) [inline shared/modules/persistence_module.go:42]: Can you update the comment on the `TreeStore`. I read it and it did not give me much context.

I'm thinking more along the lines of:

`TreeStore provides an interface that guarantees atomic operations (updates and rollbacks) made to the tree supporting the persistence context`

File: shared/modules/persistence_module.go
Comment (Olshansk) [inline shared/modules/persistence_module.go:46]: `and the ...`?

File: shared/modules/persistence_module.go
Comment (Olshansk) [inline shared/modules/persistence_module.go:47]: Can you elaborate on what this means?

File: shared/modules/persistence_module.go
Comment (Olshansk) [inline shared/modules/persistence_module.go:53]: Let's prefact it with `Debug` (e.g. DebugClearAll) and add an `IMPROVE` to move it into a separate file with the appropriate build tag

File: persistence/module.go
Comment (Olshansk) [inline persistence/module.go:40]:
```suggestion
	// txIndexer is a key-value store mapping transaction hashes to `IndexedTransaction` protos.
```

Is that right?

File: persistence/module.go
Comment (Olshansk) [inline persistence/module.go:42]:
```suggestion
	// It is needed for avoiding tx replays attacks, and some business logic around tx validation.
	// IMPORTANT: It doubles as the backing data store for the tx tree in one of the `stateTrees`.
```

File: persistence/module.go
Comment (Olshansk) [inline persistence/module.go:107]: Thoughts on having the `TreeStore` be a submodule (using dependency injection, being modular, etc...) that takes in a bus? Then you can access the `txIndexer` via `GetBus().GetPersistenceModule().GetTxIndexer()`

File: shared/modules/gomock_reflect_1203808572/prog.go
Comment (Olshansk) [inline shared/modules/gomock_reflect_1203808572/prog.go:1]: Intended to be merged in?

File: shared/modules/persistence_module.go
Comment (Olshansk) [inline shared/modules/persistence_module.go:28]: I know we didn't do a good job at this earlier on, but can make the comments more explicit on what it does and a couple words on what it's used for. It'll help future readers

File: shared/modules/persistence_module.go
Comment (Olshansk) [inline shared/modules/persistence_module.go:51]: I have a lot of context (more than most) on the work you're doing here and it was hard for my to decipher the interface and implementation. What do you think of this updated interface (note that the txi is part of the state)?

```go
// Update uses the ephemeral state in the postgres transaction pgtx at the given height and automatically
// updates all the trees backing the persistence module. 
// Updates to current or prior heights are idempotent and return the same hash every time.
// Updates to future heights returns an error.
`Update(pgtx pgx.Tx, height uint64) (string, error)`
```

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:1]: NIT: Please use `mv` in the future. 

Did you make any chances to this logic (do I need to review it) or is it just a mv?

**Review by Olshansk — CHANGES_REQUESTED** [review]
@dylanlott Do you plan on adding any unit tests to this?

File: persistence/context.go
Comment (dylanlott) [reply persistence/context.go:34]: 🙏

File: persistence/module.go
Comment (dylanlott) [reply persistence/module.go:107]: Totally open to it, is it okay if i address it in a follow up ticket?

File: shared/modules/gomock_reflect_1203808572/prog.go
Comment (dylanlott) [reply shared/modules/gomock_reflect_1203808572/prog.go:1]: Nope, I'm in the process of removing it from this PR

File: shared/modules/persistence_module.go
Comment (dylanlott) [reply shared/modules/persistence_module.go:28]: Update incoming  👍

File: shared/modules/persistence_module.go
Comment (dylanlott) [reply shared/modules/persistence_module.go:41]: Updated to be more accurate.

File: persistence/module.go
Comment (dylanlott) [reply persistence/module.go:42]: NB: The connection to the `TxIndexer` is covered in the design document for this refactor.

File: persistence/trees/trees.go
Comment (dylanlott) [reply persistence/trees/trees.go:1]: > @dylanlott Do you plan on adding any unit tests to this?

In follow up PRs, yes, I was just trying to bring it up to test parity for this PR to keep it smaller.

Comment (dylanlott) [reply persistence/trees/trees.go:1]: > NIT: Please use `mv` in the future.
> 
> Did you make any chances to this logic (do I need to review it) or is it just a mv?

Yes, you should review it all out of caution, but I attempted to keep changes minimal and separated the querying from the actual update application to make it easier to reason about.

**Comment by dylanlott** [comment]
> @dylanlott Do you plan on adding any unit tests to this?

Yes, was planning on future PRs with more test coverage. This PR was meant to bring the refactor online and achieve test parity.

File: persistence/module.go
Comment (dylanlott) [reply persistence/module.go]: Addressed in #808

File: shared/modules/persistence_module.go
Comment (Olshansk) [reply shared/modules/persistence_module.go:52]: @dylanlott Bump on this?

File: persistence/trees/trees.go
Comment (Olshansk) [reply persistence/trees/trees.go:1]: Any chances you can share a `git diff` that could help me out with it or point me to where I should review the core changes?

I understand the importance of `you should review it all out of caution` but it's 600 lines of business logic and would really help me out if if I only had to evaluate the It's 600 lines of business logic and would really

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:573]: Related to

> Yes, was planning on future PRs with more test coverage. This PR was meant to bring the refactor online and achieve test parity.

Can you add the Test function headers (w/ a TODO to implement) even if you don't plan on implementing them in this PR? That way we're not just forgetting about it.

I think that doing a git diff on this PR and getting suggestions from ChatGPT could help out.

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:409]: We should create a ticket for it if we're referencing the same techdebt item each time

File: persistence/module.go
Comment (Olshansk) [inline persistence/module.go:107]: Can you add a TODO (linking to your other PR) to make this a submodule?

**Review by Olshansk — DISMISSED** [review]

File: persistence/trees/trees.go
Comment (dylanlott) [reply persistence/trees/trees.go:408]: Created #813 👍

File: shared/modules/persistence_module.go
Comment (dylanlott) [reply shared/modules/persistence_module.go:52]: I was going to address this in #808 because it seemed better suited to handle it with the other module changes. Is that reasonable? I can handle it here if preferred.

File: persistence/trees/trees.go
Comment (dylanlott) [reply persistence/trees/trees.go:1]: https://gist.github.com/dylanlott/380848a878a72f45b8c2f15231ddfa22 Here's a gist link to the diff between state.go (previous) and trees.go (refactored). The core of the tree changes are in persistence/trees/trees.go.

The only major difference is that we moved the state.go package off of the persistence context and into its own package, and then separated the query functions from the tree update functions for readability. Otherwise, we use the same structure and approach, and especially kept the same order of trees that was declared since that's a sensitive part of the package.

**Review by h5law — COMMENTED** [review]
Looks great to me 🧙🏼‍♂️

My only concern/question is along the lines of a comment made by @Olshansk. I think I may have a slight misunderstanding on the role of the `TxIndexer` and why it is being used as the store for the transaction tree? I believed these to be separated for a reason, in my mind I seem to remember discussing this on a call before while doing the RPC spec and fixing some issues with the TxIndexer... 🤔 

Other than that the work looks great so far! I see the TODO test placeholders and if this is to be done in another PR then everything seems to be making great progress with the refactors!

**Review by Olshansk — DISMISSED** [review]
Don't merge. Testing something

**Comment by dylanlott** [comment]
> My only concern/question is along the lines of a comment made by @Olshansk. I think I may have a slight misunderstanding on the role of the `TxIndexer` and why it is being used as the store for the transaction tree?

It's the source of truth for transaction data. The tree only reflects what is persisted into the TxIndexer. If we didn't get them from the TxIndexer then we'd have to keep them in memory and pass them through from the Unit of Work module, IIRC. Instead, they're persisted to the TxIndexer.

**Comment by dylanlott** [comment]
> Other than that the work looks great so far! I see the TODO test placeholders and if this is to be done in another PR then everything seems to be making great progress with the refactors!

@h5law also yes the test placeholders will be addressed in the near future, I'm working on those while I'm spiking out the savepoints implementation

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:28]: When 👁️ ?

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:68]:
```suggestion
	// to generate the state hash.
        // TECHDEBT: Enforce this ordering in protobufs using enums

	// Actor Merkle Trees
```

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:89]:
```suggestion
// * It is responsible for atomic commit or rollback behavior
```

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:90]: I can't believe I just realized that "lazy loading" is what enables this 🤯

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:98]:
```suggestion
// Update takes an indexed transaction and a height and updates
```

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:99]:
```suggestion
// all of the trees in the treeStore affected by that transaction for that height.
```

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:121]: Can you move `sha256.New` into a local constant called `smtPathHasher` so it's not "hidden" in the code

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:116]: Optional NIT: Move the formatted string into a local var at the top of the file so it's easier to see where/how we store the data.

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:128]:
```suggestion
// TECHDEBT: Move this into a separate file with a debug build flag to avoid accidental usage in prod
func (t *treeStore) DebugClearAll() error {
```

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:104]: Can you add a comment? `NewStateTrees ...`

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:149]: Make sure to update this to the package scoped var as well

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:155]:
```suggestion
// * it returns the new state hash capturing the state of all the trees or an error if one occured
```

File: persistence/trees/trees_test.go
Comment (Olshansk) [inline persistence/trees/trees_test.go:1]: The list of headers doesn't seem to be exhaustive and I think it's a non-trivial task to figure out what to test. While you're working on this, you probably have the most context of what we need, which will be lost later.

Can you please do one of:
1. Add more test function headers on what we think we need
2. Create a new github issue (or see if one exists) and link to it here. Potentially add a deliverable to leverage ChatGPT to see what else we should test for

The lack of testing in this is one of the few things that keep me up at night so appreciate you understanding :)

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:385]: Leaving one comment but ditto elsewhere 

```suggestion
	// TECHDET(#813): Avoid this cast to int64
```

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:408]: NIT: Missing #813.

PLease check other places as well

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:415]: Given line 391, I don't think we need this

File: persistence/trees/trees.go
Comment (Olshansk) [inline persistence/trees/trees.go:475]: Prior to this change, we were using a helper:

Screenshot 2023-06-14 at 1 24 27 PM

I'm not a fan of as it lets SQL spill over into new pats of the code that were otherwise blind to it

Screenshot 2023-06-14 at 1 24 44 PM

I feel strongly that we either revert or encapsulate the logic elsewhere, or please explain the logic/reasoning here.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: persistence/trees/trees.go
Comment (dylanlott) [reply persistence/trees/trees.go:28]: Soon! Iteration 19?

File: persistence/trees/trees.go
Comment (dylanlott) [reply persistence/trees/trees.go:476]: So, there's an import cycle error that arises if we try to use the existing SQL function, which is why I had to have the SQL handling in the trees package. I think one option would be to make a sibling package in `persistence/` that abstracts away the SQL handling and then `trees` package could import it. Do you have a better idea? I didn't think it was too ugly to have the SQL handling in this package but I can understand wanting to limit that handling and scope.

Comment (Olshansk) [reply persistence/trees/trees.go:476]:
```
So, there's an import cycle error that arises if we try to use the existing SQL function, 
```

Can you show/explain what the import cycle is?

```
think one option would be to make a sibling package in persistence/ that abstracts away the SQL handling and then trees package could import it.
```

Sounds reasonable to me. Though I do want to understand the above first.


```
 I didn't think it was too ugly to have the SQL handling in this package but I can understand wanting to limit that handling and scope.
 ```
 
 Yea, I'd like to avoid this.

File: persistence/trees/trees.go
Comment (Olshansk) [reply persistence/trees/trees.go]: Cool

File: persistence/trees/trees.go
Comment (Olshansk) [reply persistence/trees/trees.go:28]: 🙌

File: persistence/trees/trees.go
Comment (h5law) [reply persistence/trees/trees.go:123]: To be clear this is actually the `TreeHasher` so maybe `smtTreeHasher` is more appropriate

File: persistence/trees/trees.go
Comment (h5law) [inline persistence/trees/trees.go:56]: Wdyt of changing the root hash calculation to be the root of an SMT with key value pairings as follows:
`stateRootTree.Update([]byte("app"), treeStore.merkleTrees["app"].Root())`

Then our state root would be `stateRootTree.Root()`

Comment (dylanlott) [reply persistence/trees/trees.go:56]: I'd like to see what this looks like spiked out but at first glance nothing jumps out at me as incompatible.

File: persistence/trees/trees.go
Comment (dylanlott) [reply persistence/trees/trees.go]: @Olshansk okay I pushed up a commit that places the functions in `persistence/sql` package and I think we should TECHDEBT refactoring the other PersistenceContext helpers that handle SQL into that packge. How does that sound?

File: persistence/sql/README.md
Comment (dylanlott) [inline persistence/sql/README.md:35]: @Olshansk the import cycle issue is described here.

File: persistence/trees/trees.go
Comment (Olshansk) [reply persistence/trees/trees.go:56]: Resolving per the ticket opened in #834. 

**Optional NIT**: Adding a TODO/techdebt in the code

File: persistence/trees/trees.go
Comment (Olshansk) [reply persistence/trees/trees.go:93]: Bump?

File: persistence/sql/README.md
Comment (Olshansk) [reply persistence/sql/README.md:35]: Screenshot 2023-06-16 at 4 00 36 PM

I'm in full support of this :)

**Review by Olshansk — APPROVED** [review]
@dylanlott I pushed a couple last nits myself (PTAL), but otherwise lgtm

File: persistence/trees/trees_test.go
Comment (dylanlott) [reply persistence/trees/trees_test.go:1]: #836 covers tests and mentions this refactor in the description. Is that sufficient here instead?

File: persistence/trees/trees.go
Comment (Olshansk) [reply persistence/trees/trees.go]: Let's do it. Wealready have 2 comments on main w.r.t to doing this

Screenshot 2023-06-16 at 4 18 32 PM

File: persistence/trees/trees.go
Comment (Olshansk) [reply persistence/trees/trees.go:92]: Bump

File: persistence/trees/trees_test.go
Comment (Olshansk) [reply persistence/trees/trees_test.go:1]: Yea, just link to the ticket in the `trees_test` file

**Review by Olshansk — CHANGES_REQUESTED** [review]
Just realied a few comments are still unresolved, so unapproving for now
Screenshot 2023-06-16 at 4 21 58 PM

File: persistence/module.go
Comment (h5law) [inline persistence/module.go:47]: Can we export these for use outside of the TreeStore module?

**Review by h5law — APPROVED** [review]
I have read through the PR and a few comments remain outstanding regarding comments and moving some things into global vars. However, I think the changes here are good. From the last few reads I cannot see anything major left to address that was not mentioned above (or in another issue).

I hope we can get this merged soon! (PS. I like the SQL separation a lot 😉)

File: persistence/trees/trees.go
Comment (h5law) [inline persistence/trees/trees.go:337]: Screenshot 2023-06-19 at 13 06 04

Can you address these linter errors? Either bring the helper functions into the SQL package or remove them

File: shared/modules/persistence_module.go
Comment (dylanlott) [reply shared/modules/persistence_module.go]: I updated this comment to more succinctly reflect the situation regarding pgx and transaction isolation.

File: persistence/trees/trees.go
Comment (dylanlott) [reply persistence/trees/trees.go:115]: Updated to `var` in local scope. It can't be a constant because of the hash.Hash type.

File: persistence/trees/trees.go
Comment (dylanlott) [reply persistence/trees/trees.go:92]: It doesn't take an indexed transaction, though, it takes a transaction indexer. Updated to be more specific, though.

File: persistence/trees/trees.go
Comment (dylanlott) [reply persistence/trees/trees.go]: Is there an existing issue for this? If not, I'll make one for it. Otherwise, I'll consider these comments sufficient.

File: persistence/trees/trees_test.go
Comment (Olshansk) [inline persistence/trees/trees_test.go:5]:
```suggestion
// TECHDEBT(#836): Tests added in https://github.com/pokt-network/pocket/pull/836
```

I'm just thinking about automation and regex parsing

**Review by Olshansk — APPROVED** [review]
Let's get this over the finish line!

**Pull request merged**
